### ISSUE TITLE
sokol_nuklear zig bindings

### ISSUE THREAD

**Issue by chakany** [floooh/sokol-zig#73]
Hello!
I was looking to add sokol_nuklear bindings to this repository for usage inside of my project, but I noticed that the bindings were all machine-generated. I couldn't find a tool anywhere that would let me generate these bindings, but I did see your blog post.

Could you shed some light on how I can make these bindings please?
Thank you

**Comment by floooh** [comment]
The important stuff is here:

https://github.com/floooh/sokol/tree/master/bindgen

(I just notice that the readme is incomplete, you'll also need to clone the new D bindings into the bindgen directory: `git clone https://github.com/floooh/sokol-d`

I think it's best to do roughly the same as sokol_imgui.h (which is a bit special in that there's no example in sokol-zig, because it would require a cimgui+Dear ImGui dependency, instead the example is in a separate repository:

https://github.com/floooh/sokol-zig-imgui-sample

If you get stuck, let me know. Also if you get it working let me also know because then we can add it to the official bindings ;)

**Comment by floooh** [comment]
Basically:

- add a new line here, similar to the sokol_imgui.h entry: https://github.com/floooh/sokol/blob/7b20c1936229370277d1c61bde950bce194de584/bindgen/gen_all.py#L31
- add a new line here: https://github.com/floooh/sokol/blob/7b20c1936229370277d1c61bde950bce194de584/bindgen/gen_zig.py#L25
- ...and here: https://github.com/floooh/sokol/blob/7b20c1936229370277d1c61bde950bce194de584/bindgen/gen_zig.py#L39
- ...in the cloned sokol-zig subdirectory, add a sokol_nuklear.c file here, with content similar to sokol_imgui.h, don't worry about the Nuklear header dependency yet, this is only needed for building the sokol C library later, not for generating the bindings

Back in the bindgen directory, try running `python3 gen_all.py` and look for errors. If it works, the sokol_nuklear.h header should have been copied to `sokol-zig/src/sokol/c/`, and a file `sokol-zig/src/sokol/nuklear.zig should have been created.

**Comment by chakany** [comment]
Hello, I did the following steps, and I got something simpler than I imagined.
```zig
// machine generated, do not edit

const builtin = @import("builtin");
const sg = @import("gfx.zig");
const sapp = @import("app.zig");

// helper function to convert a C string to a Zig string slice
fn cStrToZig(c_str: [*c]const u8) [:0]const u8 {
    return @import("std").mem.span(c_str);
}
```
So, none of the actual bindings are there. Very strange.

**Comment by floooh** [comment]
Hmm, I get a different problem which I didn't anticipate:

```
In file included from sokol-zig/src/sokol/c/sokol_nuklear.c:6:
sokol-zig/src/sokol/c/sokol_nuklear.h:322:2: error: "Please include nuklear.h before sokol_nuklear.h"
#error "Please include nuklear.h before sokol_nuklear.h"
 ^
sokol-zig/src/sokol/c/sokol_nuklear.h:440:24: error: unknown type name 'nk_handle'
SOKOL_NUKLEAR_API_DECL nk_handle snk_nkhandle(snk_image_t img);
                       ^
sokol-zig/src/sokol/c/sokol_nuklear.h:441:60: error: unknown type name 'nk_handle'
SOKOL_NUKLEAR_API_DECL snk_image_t snk_image_from_nkhandle(nk_handle handle);
```

Problem is that sokol_nuklear.h needs the nuklear.h header even for the declarations, because of the use of nk_handle here:

https://github.com/floooh/sokol/blob/7b20c1936229370277d1c61bde950bce194de584/util/sokol_nuklear.h#L440

...this is actually tricky to solve because requiring the nuklear.h header for bindings generation is a no-go.

I don't have a quick solution to offer, we basically need to get rid somehow of all nuklear.h dependencies (like nk_handle). Maybe with a preprocessor define somewhere...

**Comment by chakany** [comment]
> I don't have a quick solution to offer, we basically need to get rid somehow of all nuklear.h dependencies (like nk_handle). Maybe with a preprocessor define somewhere...

would it be possible to activate the preprocessor `if`s from the command line at compile time? and only when building for bindings?

**Comment by floooh** [comment]
Yep, cleanest solution would be to add a define in the clang call here:

https://github.com/floooh/sokol/blob/7b20c1936229370277d1c61bde950bce194de584/bindgen/gen_ir.py#L102

...maybe this works:

`cmd = ['clang', '-Xclang', '-ast-dump=json', '-D', 'SOKOL_BINDGEN', '-c' ]`

...and then in the sokol_nuklear.c file we can check for `SOKOL_BINDGEN` and provide the required Nuklear types directly before sokol_nuklear.h is included, e.g.:

```
#if defined SOKOL_BINDGEN
typedef union {void *ptr; int id;} nk_handle;
#endif
```

**Comment by chakany** [comment]
if the nuklear types have a `snuklear_` prefix instead of a `nk_` prefix that would work right?
after that, well it's a matter of redefining the `nk_` type to `snk_` if !`SOKOL_BINDGEN`.

**Comment by chakany** [comment]
been working on this for a little while, i have some basic types.
```c
#if defined(IMPL)
#define SOKOL_NUKLEAR_IMPL
#define NK_IMPLEMENTATION
#include "nuklear.h"
#endif
#include "sokol_defines.h"
#include "sokol_app.h"
#include "sokol_gfx.h"
#ifdef SOKOL_BINDGEN
struct nk_context;
typedef union {void *ptr; int id;} nk_handle;
typedef unsigned int nk_flags;
typedef int(*nk_plugin_filter)(const struct nk_context*, nk_handle, int*, int);
#endif
#include "sokol_nuklear.h"
```

generations go until it reaches a `struct nk_context *` type. it outputs
```py
Error as_c_arg_type(): struct nk_context *
```

I believe that the `nk_context` struct is not being included in the AST dump, and that's why it cannot be found. So, i'm trying to get the `nk_` prefix included in the `gen_zig.py`/`gen_ir.py` scripts, and I think i'm stuck.

Thanks again.

**Comment by DylanLukes** [comment]
Hi @chakany, do you have a work-in-progress published somewhere? I'm interested in hacking on this a bit myself as well.

– Dylan

**Comment by DylanLukes** [comment]
Ah... looks like `nk_context` is used inside `_snk_state_t`[1], so its layout actually has to be known. That would seem to preclude the usual opaque pointer tricks without some more significant rewriting of `sokol_nuklear.h`.

[1] As is `nk_font_atlas` for that matter...
